lib/sampler/storage/src/lib/span-dao.js

```diff
@@ -0,0 +1,216 @@
+      update.$set.end_dedup_id = dedupID;
+    const options = {};
+    try {
+      const status = await this.collection.findOneAndUpdate(filter, update, options);
+      return !!status.value;
+    } catch (e) {
+      if (e.code !== MONGO_DUPLICATE_KEY_ERROR_CODE)
+        throw e;
+      return false;
+    }
+  }
+
+  async deleteSpansByIDs(ids) {
+    await this.collection.deleteMany({
+      '_id': {
+        '$in': ids
+      }
+    });
+  }
+
+  async getSpanByTarget(target) {
+    const cursor = this.collection.find({
+      'target': target
+    }).limit(1);
+    if (!await cursor.hasNext())
+      return undefined;
+    return await cursor.next();
+  }
+
+  async existsSpanWithStartDedupID(id) {
+    const cursor = this.collection.find({
+      'start_dedup_id': id
+    }).limit(1);
+    return await cursor.hasNext();
+  }
+
+  async existsSpanWithEndDedupID(id) {
+    const cursor = this.collection.find({
+      'end_dedup_id': id
+    }).limit(1);
+    return await cursor.hasNext();
+  }
+
+  async updateSpanPlannedInterval(id, interval, version) {
+    const filter = {
+      '_id': id,
+      'processing.version': version
+    };
+    const update = {
+      '$set': {
+        'processing.planned_interval': interval,
+        'processing.last_change_at': moment.now(),
+        'processing.version': version + 1
+      }
+    };
+    const options = {};
+    const status = await this.collection.findOneAndUpdate(filter, update, options);
+    return !!status.value;
+  }
+
+  async updateSpanProcessedInterval(id, interval, complete, version) {
+    const filter = {
+      '_id': id,
+      'processing.version': version
+    };
+    const update = {
+      '$set': {
+        'processing.complete': complete,
+        'processing.last_interval': interval,
+        'processing.planned_interval': null,
+        'processing.last_change_at': moment.now(),
+        'processing.version': version + 1
+      }
+    };
+    const options = {};
+    const status = await this.collection.findOneAndUpdate(filter, update, options);
+    return !!status.value;
+  }
+
+  async findIncompleteSpans(before, offset, limit) {
+    const cursor = this.collection.find({
+      'processing.complete': false,
+      'processing.last_interval.end': {
+        '$lt': before
+      }
+    }).skip(offset).limit(limit);
+    return await cursor.toArray();
+  }
+
+  async findCompleteSpans(before, offset, limit) {
+    const cursor = this.collection.find({
+      'processing.complete': true,
+      'processing.last_interval.end': {
+        '$lt': before
+      }
+    }).skip(offset).limit(limit);
+    return await cursor.toArray();
+  }
+}
+
+const createTargetIndex = async (collection) => {
+  const definition = {
+    'target.organization_id': 1,
+    'target.space_id': 1,
+    'target.consumer_id': 1,
+    'target.resource_id': 1,
+    'target.plan_id': 1,
+    'target.resource_instance_id': 1,
+    'target.correlation_id': 1
+  };
+  const options = {
```

Comment:
Might be worth mentioning/commenting that options cannot be updated without dropping Index as per https://docs.mongodb.com/manual/reference/method/db.collection.createIndex/#behaviors

Reply:
I doubt we would need to change the options. Removing or adding unique constrants without serios consideration is asking for trouble. In general, those indices we have designed to be `unique` are meant to be `unique`, otherwise the whole algorithm will fail.

In general, we discussed that the collection and index creation could be moved into a separate task that is performed prior to deployment. In the future, we might want to do some updates, that may require that. However, for the first iteration this would be an overkill. It is complex to develop, integrate with CF deployment flows (since tasks require running applications) and any usage of the DAO without prior creation of the indices would be a major problem, which the current implementation prevents (as DAO usage ensures creation of prerequisites).

Reply:
We have added a comment in the file to indicate that.

Reply:
We added a note regarding this as a comment in the code.